rc/extra/gdb.kak

```diff
@@ -0,0 +1,282 @@
+                    # QUOTE => \\QUOTE
+                    gsub("'\''", "\\\\'\''", print_value)
+                    # eval -client $client QUOTE info  -- \QUOTE $string \QUOTE QUOTE
+                    send("eval -client '"$kak_client"' '\''info -- \\'\''" print_value "\\'\'\''")
+                    printing = 0
+                }
+            }
+            add-highlighter ref gdb
+            '
+        } 2>/dev/null >/dev/null &
+        printf "$!" > "${tmpdir}/pid"
+        printf "set-option global gdb_dir %s\n" "$tmpdir"
+        printf "info \"%s\"\n" "Please instruct gdb to \\\"new-ui mi3 ${tmpdir}/pty\\\""
+    }
+    hook -group gdb global BufOpenFile .* %{
+        gdb-refresh-location-flag
+        gdb-refresh-breakpoints-flags %val{buffile}
+    }
+    hook -group gdb global KakEnd .* %{
+        gdb-stop
+    }
+}
+
+define-command gdb-stop %{
+    try %{
+        %sh{
+            if [ -n "$kak_opt_gdb_dir" ]; then
+                pkill -P $(cat "${kak_opt_gdb_dir}/pid")
+                rm "${kak_opt_gdb_dir}/pid" "${kak_opt_gdb_dir}/pipe"
+                rmdir "$kak_opt_gdb_dir"
+            else
+                printf %s\\n raise
+            fi
+        }
+        set-option global gdb_dir ""
+
+        gdb-clear-location
+        gdb-clear-breakpoints
+
+        remove-hooks global gdb
+    }
+}
+
+define-command gdb-jump-to-location %{
+    %sh{
+        if [ -n "$kak_opt_gdb_location_info" ]; then
+            line=$(printf %s "$kak_opt_gdb_location_info" | cut -d \| -f 1)
+            buffer=$(printf %s "$kak_opt_gdb_location_info" | cut -d \| -f 2-)
+            printf "edit -existing \"%s\" %s\n" "$buffer" "$line"
+        fi
+    }
+}
+
+define-command -params 1 gdb-cmd %{
+    %sh{
+        if [ -n "$kak_opt_gdb_dir" ]; then
+            printf %s\\n "$1" > "$kak_opt_gdb_dir"/pipe
+        fi
+    }
+}
+
+define-command gdb-run              %{ gdb-cmd run }
+define-command gdb-start            %{ gdb-cmd start }
+define-command gdb-step             %{ gdb-cmd step }
+define-command gdb-next             %{ gdb-cmd next }
+define-command gdb-finish           %{ gdb-cmd finish }
+define-command gdb-continue         %{ gdb-cmd continue }
+define-command gdb-advance          %{ gdb-cmd "advance %val{buffile}:%val{cursor_line}" }
+define-command gdb-set-breakpoint   %{ gdb-cmd "break %val{buffile}:%val{cursor_line}" }
+define-command gdb-clear-breakpoint %{ gdb-cmd "clear %val{buffile}:%val{cursor_line}" }
+define-command gdb-toggle-breakpoint %{
+    try %{
+        %sh{
+            printf %s\\n "$kak_opt_gdb_breakpoints_info" | tr ':' '\n' |
+            while read -r current; do
+                buffer=$(printf %s "$current" | cut -d \| -f 4-)
+                line=$(printf %s "$current" | cut -d \| -f 3)
+                if [ "$buffer" = "$kak_buffile" -a "$line" = "$kak_cursor_line" ]; then
+                    printf %s\\n raise
```

Comment:
Since the while loop is in a subshell I wouldn't be able to `exit` without also calling `set-breakpoint`. I don't think a heredoc solution like I did in `gdb-handle-breakpoint-deleted` would be cleaner, so unless you have a simpler solution to this I'll leave it as is.